log-warper-core/src/Log/Core/Class.hs

```diff
@@ -0,0 +1,201 @@
+
+    -- | Runs given monadic computation with changed 'LogAction'.
+    withLogAction :: (LogAction m r -> LogAction m r) -> m a -> m a
+
+-- | Helper function. Helps to create instance of 'MonadLogger', especially to
+-- implement 'withLogAction' method.
+unliftLogAction :: (MonadTrans t, Monad m)
+                => (t m () -> m a)  -- ^ Transformer unwrapper for logging action
+                -> (LogAction (t m) r -> LogAction (t m) r)
+                -> LogAction m r
+                -> LogAction m r
+unliftLogAction unlift f (LogAction action) = case f $ LogAction (lift . action) of
+    LogAction transAction -> LogAction (void . unlift . transAction)
+
+instance MonadLogger m r => MonadLogger (IdentityT m) r where
+    withLogAction :: (LogAction (IdentityT m) r -> LogAction (IdentityT m) r) -> IdentityT m a -> IdentityT m a
+    withLogAction = mapIdentityT . withLogAction . unliftLogAction runIdentityT
+
+instance MonadLogger m r => MonadLogger (MaybeT m) r where
+    withLogAction :: (LogAction (MaybeT m) r -> LogAction (MaybeT m) r) -> MaybeT m a -> MaybeT m a
+    withLogAction = mapMaybeT . withLogAction . unliftLogAction runMaybeT
+
+instance MonadLogger m r => MonadLogger (ExceptT e m) r where
+    withLogAction :: (LogAction (ExceptT e m) r -> LogAction (ExceptT e m) r) -> ExceptT e m a -> ExceptT e m a
+    withLogAction = mapExceptT . withLogAction . unliftLogAction runExceptT
+
+instance MonadLogger m r => MonadLogger (ReaderT e m) r where
+    withLogAction :: (LogAction (ReaderT e m) r -> LogAction (ReaderT e m) r) -> ReaderT e m a -> ReaderT e m a
+    withLogAction f action = do
+        env <- ask
+        mapReaderT (withLogAction (unliftLogAction (flip runReaderT env) f)) action
+
+-- | __TODO:__ Not sure about this instance; it might contain error
+instance MonadLogger m r => MonadLogger (StateT s m) r where
+    withLogAction :: (LogAction (StateT s m) r -> LogAction (StateT s m) r) -> StateT s m a -> StateT s m a
+    withLogAction f action = do
+        st <- get
+        mapStateT (withLogAction (unliftLogAction (flip runStateT st) f)) action
+
+{- | Perform logging action with given logging record @r@. You can use this
+function like this:
+
+@
+__foo__ :: 'MonadLogger' m 'String' => m ()
+__foo__ = __do__
+    'logR' "Some logging"
+    'logR' "Another logging"
+@
+-}
+logR :: MonadLogger m r => r -> m ()
+logR r = do
+    LogAction action <- askLogAction
+    action r
+
+{- | Executes given monadic logging action by applying function to every logging record.
+
+@
+__app__ :: 'MonadLogger' m 'String' => m ()
+__app__ = 'withR' ("app:" ++) $ __do__
+    'logR' "Some logging"
+    'logR' "Another logging"
+@
+-}
+withR :: MonadLogger m r => (r -> r) -> m a -> m a
+withR = withLogAction . cmap
+
+{- | Simple implementation for 'MonadLogger' type class. This data type just
+stores 'LogAction' which performs logging action under monad @m@.
+-}
+newtype ActionT r m a = ActionT
```

Comment:
Now you're just copying `logging-effect` but with worse naming, aren't you?

Reply:
@int-index I never heard about `logging-effect` library before this comment, my bad. I've implemented these data types and type classes using my poor experience and your suggestions about monoidal actions. And I always appreciate better suggestions for naming!

Reply:
Maybe just take `logging-effect` as the basis? The definitions are very similar, and we shouldn't fracture the ecosystem.

Reply:
@int-index @volhovm Okay, so I looked at `logging-effect` library. Approaches looks similar. But I still like my approach more :stuck_out_tongue_winking_eye:

`LoggingT` looks like better name to me as well. But I've already used `LoggerT` name inside `log-warper`. I found having both `LoggingT` and `LoggerT` confusing. So both of them should be renamed.

Here the reasons why I like my approach more and why I don't want to depend on `logging-effect`:

1. `logging-effect` has a huge number of dependencies we don't really need. If maintainer could move type class to something like `logging-effect-core` we could think on using this package as a basis for `log-warper`.
2. `MonadLog` type class from `logging-effect` looks much more complicated to me than `MonadLogger` from `log-warper-core`. I don't really understand this definition:

```haskell
class Monad m => MonadLog message m | m -> message where
    logMessageFree :: (forall n. Monoid n => (message -> n) -> n) -> m ()
```

  Well, what I understand perfectly is that I'm not a very smart developer. So maybe this is just my problem. But I'm also not the most stupid one. And from my experience I can tell that things that are easier to understand and grasp are also easier to reason about and it's easier to find bugs and implement instances and so on.

3. `Handler` in `logging-effect` is a type alias. `LogAction` is `log-warper-core` is a `newtype`. I like `newtypes` more. Type alias has its own benefits. But I like the following benefits of `newtype` more:

    * Ability to implement custom (and, as a consequence, more efficient) `Semigroup` and `Monoid` instances.
    * Type-safety? Anyone?..

4. Types of combinators for modifying logging action are fixed to specific `LoggingT` data type while in `log-warper-core`  they work with monad itself. Compare:

    * `logging-effect`:
        ```haskell
        mapLogMessageM :: MonadLog message' m 
                       => (message -> m message') 
                       -> LoggingT message m a 
                       -> m a
        ```
    * `log-warper-core`:
        ```haskell
        ward :: Applicative m => (a -> Bool) -> LogAction m a -> LogAction m a
        cbind :: Monad m => (a -> m b) -> LogAction m b -> LogAction m a
        withLogAction :: (LogAction m r -> LogAction m r) -> m a -> m a
        ```

  If I add more nested `mapLogMessageM` calls in `logging-effect` performance will be decreased because the more transformers you have on stack ( (and the bigger height of transoferms tower) the lower performance will be. With `log-warper-core` I don't add extra new transformers, I only apply functions to actions and I only change action which is stored inside monadic context.

5. I wrote big documentation and I don't want to throw it out...